File: plot/src/main/java/net/fwitz/math/plot/renderer/palette/Palette.java
package net.fwitz.math.plot.renderer.palette;

import java.awt.*;

public interface Palette {
    /** The number of colors included in this palette. */
    int size();

    /**
     * Selects a color by index, without wrapping.
     *
     * @param index the color index, which must be in the range {@code [0, size)}; that is, it must be non-negative
     *              and less than {@link #size()}.
     * @return the selected color
     * @throws IndexOutOfBoundsException if the supplied index is out of range
     */
    Color index(int index);

    /**
     * Returns a color from the palette, using all available colors and wrapping as needed.
     *
     * @param index the (non-negative) color index
     * @return the corresponding color
     */
    default Color indexAny(int index) {
        return index(index % size());
    }

    /**
     * Returns a color from the palette, using all available colors except the one at index {@code 0} and wrapping
     * as needed.
     *
     * @param index the (non-negative) color index
     * @return the corresponding color
     */
    default Color indexExcluding0(int index) {
        int max = size() - 1;
        int n = index % max;
        return index((n != 0) ? n : max);
    }

    /**
     * Returns a color constructed from the palette by interpolating between the color that would be selected
     * by {@link #indexAny(int)} using {@link Math#floor(double) floor(x)} and the next color in the palette,
     * wrapping back around to index {@code 0} if necessary.
     *
     * @param x the color index, with the integer portion representing the first color's index as per
     * {@link #index(int)} and the mantissa indicating the amount of the next color that should be mixed in
     * @return the interpolated color value
     */
    default Color interpolateAny(double x) {
        int n = ((int) Math.floor(x));
        double mantissa = x - n;
        n %= size();
        int nPlus1 = n + 1;

        return Interpolate.interpolate(
                index(n),
                index((nPlus1 != size()) ? nPlus1 : 0),
                mantissa);
    }

    /**
     * Returns a color constructed from the palette by interpolating between the color that would be selected
     * by {@link #indexExcluding0(int)} (int)} using {@link Math#floor(double) floor(x)} and the next color
     * in the palette, wrapping back around to index {@code 1} if necessary.
     *
     * @param x the color index, with the integer portion representing the first color's index as per
     * {@link #index(int)} and the mantissa indicating the amount of the next color that should be mixed in
     * @return the interpolated color value
     */
    default Color interpolateExcluding0(double x) {
        int max = size() - 1;
        int n = ((int) Math.floor(x));
        double mantissa = x - n;
        n %= max;

        return Interpolate.interpolate(
                index((n > 0) ? n : max),
                index(n + 1),
                mantissa);
    }
}

File: plot/src/main/kotlin/net/fwitz/math/plot/renderer/palette/Palette.kt
package net.fwitz.math.plot.renderer.palette

import java.awt.Color
import kotlin.math.floor

interface Palette {
    /** The number of colors included in this palette.  */
    val size: Int

    /**
     * Selects a color by index, without wrapping.
     *
     * @param index the color index, which must be in the range `[0, size)`; that is, it must be non-negative
     * and less than [.size].
     * @return the selected color
     * @throws IndexOutOfBoundsException if the supplied index is out of range
     */
    operator fun get(index: Int): Color

    /**
     * Returns a color from the palette, using all available colors and wrapping as needed.
     *
     * @param index the (non-negative) color index
     * @return the corresponding color
     */
    fun indexAny(index: Int) = get(index % size)

    /**
     * Returns a color from the palette, using all available colors except the one at index `0` and wrapping
     * as needed.
     *
     * @param index the (non-negative) color index
     * @return the corresponding color
     */
    fun indexExcluding0(index: Int): Color {
        val max = size - 1
        val n = index % max
        return get(if (n != 0) n else max)
    }

    /**
     * Returns a color constructed from the palette by interpolating between the color that would be selected
     * by [.indexAny] using [floor(x)][Math.floor] and the next color in the palette,
     * wrapping back around to index `0` if necessary.
     *
     * @param x the color index, with the integer portion representing the first color's index as per
     * [.index] and the mantissa indicating the amount of the next color that should be mixed in
     * @return the interpolated color value
     */
    fun interpolateAny(x: Double): Color {
        var n = floor(x).toInt()
        val mantissa = x - n
        n %= size
        val nPlus1 = n + 1
        return Interpolate.interpolate(
            get(n),
            get(if (nPlus1 != size) nPlus1 else 0),
            mantissa
        )
    }

    /**
     * Returns a color constructed from the palette by interpolating between the color that would be selected
     * by [.indexExcluding0] (int)} using [floor(x)][Math.floor] and the next color
     * in the palette, wrapping back around to index `1` if necessary.
     *
     * @param x the color index, with the integer portion representing the first color's index as per
     * [.index] and the mantissa indicating the amount of the next color that should be mixed in
     * @return the interpolated color value
     */
    fun interpolateExcluding0(x: Double): Color? {
        val max = size - 1
        var n = floor(x).toInt()
        val mantissa = x - n
        n %= max
        return Interpolate.interpolate(
            get(if (n > 0) n else max),
            get(n + 1),
            mantissa
        )
    }
}